[DON'T MERGE] Dummy pull request to allow feedback on what's currently in master

madlep opened the pull request:
Created this by running

``` sh
git branch madlep_feedback 2bc787f10e8bd7d8e533230f9ef9d5b1e18d1f40
git checkout madlep_feedback
git push origin madlep_feedback
```

This resets this branch back to the start of the project.

Then created pull request to merge master into this branch.

So it shows the full project as being the diff that will be applied, and feedback can be done on the whole thing.

Commit:
initial commit

Commit:
Merge branch 'master' of github.com:rawdamedia/quoil

Commit:
ReadMe.md

Commit:
basic command line parsing

—help and —interval options being parsed

Commit:
more tests

Commit:
use defaults.exs for storing default values

Commit:
help request takes precedence over other switches

Commit:
log_file_name made optional

Commit:
add switch for --number

Commit:
changed the return tuple from fn process_switches

this will make it easier to add other options in the future, such as to
keep the detailed ping statistics…

Commit:
list of switches now a map

Commit:
more tests and documentation

Commit:
minor corrections

Commit:
doc string update

Commit:
docs + run_ping function

Commit:
parse_results

basic infrastructure

Commit:
parse_result: targetURL & targetIP

Commit:
parse_result: packet numbers

Commit:
parse_result: stats

Commit:
write_log to console

Commit:
write_log to file

Commit:
script packaged

Commit:
Extract ArgsProcessor as separate module

Commit:
Extract LogResults as separate module

Commit:
make parse_result function dry

Commit:
Extract parse_result into new module

Commit:
Extract parse_result to new module

Commit:
finalise 0.2

Commit:
prepare for repeater functionality

Commit:
more preparation for repeater functionality

Commit:
Quoil.Repeat implemented fully

Commit:
README updated

Instructions on how to use the precompiled `quoil` command-line utility

Commit:
README updated

Commit:
README updated

Commit:
BUGFIX

failing to exit if repeat not set until wait period expires

Commit:
BUGFIX

failing to exit if repeat not set until wait period expires

Commit:
BUGFIX

calling quoil without any arguments did not return :help

Commit:
BUGFIX

calling quoil without any arguments did not return :help

Commit:
updated version number

Commit:
updated version number

Commit:
Covered for test coverage analysis

Commit:
Covered for test coverage analysis

Commit:
minor display improvements

Commit:
minor display improvements

madlep commented:
Is `Elixir.Quoil.CLI.beam` needed to be checked in? Usually would avoid binary files. `quoil` makes sense I guess, but not sure if that one does.

madlep commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+/_build
```
I'd ignore the `cover` directory too, as it's an artefact generated by tests, and not something that needs to be distributed.

rawdamedia replied:
#### Ignoring `/_build`

I actually considered doing this, but I wanted to make it easier for myself and others to use the precompiled components on systems that don't have Elixir.  My main concern was to be able to run this from my NAS that has an easy way to install Erlang but not Elixir.  I presume, though, that once I run `mix escript.build`, then all the required bits are packaged inside the executable, which I can download directly from Github onto my NAS (?) and therefore should probably exclude the `/_build` directory.

I guess if I ever want to use just part of an app, it is far better to import that functionality into the source code, rather than trying to work with precompiled beam files...

#### Including test coverage analysis in repository

I included this as a way of trying to be helpful to others quickly scanning the repository - hoping that the fact that I care about tests and test coverage would be seen as a positive... ??

madlep replied:
I'm not too familiar with how escript works, but I think that you're right: it should just need the escript file, and Erlang installed in the target environment.

I was just curious about the `beam` file in the top directory, as it seems out of place, when most of the compiled binaries would be in `_build`

Makes sense to have something runnable available to users without them having to build it themselves - especially escript - which is pretty portable so long as Erlang is available.

As for coverage - it's definitely a positive that it's there, but normally it wouldn't be checked it, as it can be generated easily again when needed, and makes the repository larger than it needs to be.

madlep commented on README.md:
```diff
@@ -0,0 +1,39 @@
+
+* *\-\-number nr* sets the number of pings in each set used to generate the ping statistics.  If it is not specified, it defaults to *:number_of_pings* specified in [**config/config.exs**](config/config.exs) (*\-\-number* can be shortened to *-n*).
+
+* *\-\-repeat nr* sets the number of times that the ping sets are to be repeated.  If it is not specified, it defaults to *nil* (*\-\-repeat* can be shortened to *-r*)
```
what does `nil` mean in this context? I'm assuming `0`? Better to be more explicit in the user facing documentation about what the effect of a config option is

madlep commented on lib/quoil/cli.ex:
```diff
@@ -0,0 +1,61 @@
+defmodule Quoil.CLI do
```
the responsibilities for `Quoil.CLI` seem confused.
- it's the entry point to the app
- it handles console output for help, and general error handling of bad args
- it runs the main business logic of the app.

Focus on each unit of code having one responsibility.

It probably makes sense to turn this into a module whose responsibility is delegating to other parts of the app
- move help/error handling and early termination into `Quoil.ArgsProcessor`
- create a new `Ping` module that contains the main business logic, and move the body of `run_ping` there

rawdamedia replied:
I recognised this as the major problem, but I simply ran out of time because I needed to get some data to back my conclusions when speaking to the Telstra techs about my internet problems.

### `Ping` Module

Creating a new `Ping` module was number 1 on my TO-DO list... mainly because I wanted to add a way of implementing the ping functionality without needing to run shell commands - please see my questions on your suggestion on this matter below.

### `terminate_early?` function

I wasn't happy about cluttering up `Quoil.CLI` with the `terminate_early?` function and all the error handling either, but I was a little worried about hiding it in a module - I was sort of thinking that by having something like that as part of the chain, then it makes it clear that _this is a point where you could have early termination_... maybe I am overthinking it, but I thought that since this is a major branch point for the program, it might be helpful to somehow highlight it - **should I be even worried about this when writing more complex code?**

(I guess I could just have a comment after the call to `parse_args`, but I thought the idea was to have self-commenting code... once again, probably overthinking it, and I just need the sort of perspective that comes from years of development work ;) )

madlep replied:
Understandable about getting code working. Useful code is better than perfect code that is never run :smile: 

I think that it is clear to move the termination logic into `parse_args` and comment it saying that it may terminate the program. The main thing is to make code obvious and unsurprising. To me anyway, having `parse_args` complain and blow up on bad args is fine and makes a lot of sense. Those kind of calls can get quite subjective based on personal preference though, and there isn't any right answer.

Alternatively, you could go for a ["single function exit"](https://stackoverflow.com/questions/4838828/why-should-a-function-have-only-one-exit-point) style (another subjective opinion...) and have something that looks like this:

```elixir
case Quoil.ArgsProcessor.parse_args do
  {:ok, ip_to_ping, switches, log_file_name} ->
    run_ping(ip_to_ping, switches, log_file_name)
    |>  ...
  {:info, message} ->
    IO.puts(:stdio, message)
    System.halt(0)
  {:error, message} ->
    IO.puts(:stderr, message)
    System.halt(1)
end
```

So with that approach, `parse_args` returns a tuple that can be matched on. The nice thing about this is that `parse_args` is responsible for handling all the messages, and logic, and the main app just carries on, or exits depending on it.

Note that the error case is outputting to stderr rather than stdin. Also, it's exiting with status 1 as opposed to using `:abort`. `:abort` will cause a core dump and is meant for unexpected crashes and other fatal errors, rather than bad user input.

madlep commented on lib/quoil/cli.ex:
```diff
@@ -0,0 +1,61 @@
+    System.halt(0)
+  end
+  def terminate_early?(:error) do
+    IO.puts "ERROR: There was an error with the command line switches."
```
a pattern that can be helpful for handling errors in a chained block like this is to use an [error monad](http://zohaib.me/monads-in-elixir-2/#error)

That way you don't have to explicitly have error handling code in the middle of your main logic.

rawdamedia replied:
I am only up to Prolog in Bruce Tate's "Seven Languages in Seven Weeks"... but I can't wait to get to Haskell !!!  

The blog post you linked to talks about two different libraries - Monad and MonadEx.  Just wondering:
- What benefits I get from using them on top of just passing around a Tuple where the first element is either _:ok_, _:error_ or _:help_?  
- In the example given, the error is generated by the last function in the chain, but what happens if the first function generates the error?  Does using these libraries give me some automatic bypassing of subsequent functions?  ...or will my_sum blow up if it is passed an error monad containing an error instead of a value?

madlep replied:
Using an error monad library means your functions don't see the `:ok` or `:error` element in the tuple, and just the payload. The monad effectively wraps metadata around the actual data you are working with, and handles wrapping, and unwrapping it.

In the case of an error, it **doesn't** call subsequent functions. It just returns the error straight away and bypasses them.

It's a more elegant way of writing deep nested logic to handle errors at any point.

So rather than writing

```elixir
case do_a_thing(data) do
  {:ok, payload} ->
    case do_another_thing(payload) do
      {:ok, another_payload} ->
        case do_yet_another_thing(another_payload) do
          {:ok, yet_another_payload} -> yet_another_payload
          {:error, error} -> handle_error(error)
        end
      {:error, error} -> handle_error(error)
    end
  {:error, error} -> handle error
end
```

you can do
```elixir
result = Error.p do
  Error.return(data)
  |> do_a_thing
  |> do_another_thing
  |> do_yet_another_thing
end

case result do
  {:error, error} -> handle_error(error)
  {:ok, value} -> value
end
```

It's less commonly used in Elixir than in something like Haskell, but it makes code like this much cleaner and easier to work with

madlep commented on lib/quoil/cli.ex:
```diff
@@ -0,0 +1,61 @@
+  """
+  def run_ping({ip_to_ping, switches, log_file_name}) do
+    args = ["-q", "-c", to_string(switches[:number]), "-W", to_string(switches[:interval]), ip_to_ping]
+    {rslt, exit_code} = System.cmd("ping", args)
```
rather than shelling out to a separate OS process to run `ping`, have a look at https://github.com/msantos/gen_icmp

You'll need to add logic to handling the counts, as that will only send one ping.

But it keeps the logic implemented in Erlang, and it returns a structure already parsed that you can use further. It also removes the overhead of calling a system process.

rawdamedia replied:
Thanks for that pointer - looks like exactly what I need!

One concern I have about using this library over the system call to `ping` is that in gen_icmp it talks about the fact that admin privileges may be required in order to open a socket, and this is not the case for ping.  Might make people worry if a utility that is essentially only supposed to be sending a series of pings is asking for administrative privileges...! (I personally would be worried if this was someone else's app and I hadn't read the code carefully)

Perhaps I will keep the current functionality as a fall-back in case the privileges aren't adequate...

rawdamedia replied:
I presume that, given network communication is one of the major use-cases for Elixir and Erlang, someone would have to deliberately restrict the language from making socket connections without admin privileges...?

madlep replied:
Not sure how it would work, you'd have to test it.

But I'm guessing that the privileges required for a user to execute the `ping` command would be the same as those required to execute an ICMP (ping) request from elixir code. So the user would need to set them up either way.

The executable itself may need different permissions set though. Not sure sorry.

rawdamedia replied:
Unfortunately my concern has been confirmed (at least on my mac)

In order to get gen_icmp installed properly I had to run `sudo mix deps.compile` as just doing `mix deps.get` did not generate the dynamic library for _procket_.  

Opening up a socket seems to work in non-privileged mode (a PID is returned) but sending pings through it returns:
> [{:error, :timeout, 'telstra.com.au', {58, 162, 89, 137}}, {:error, :timeout, 'www.google.com.au', {58, 162, 62, 84}}]

If this is run via `sudo` it works properly:
> [{:ok, 'telstra.com.au', {58, 162, 89, 137}, {58, 162, 89, 137}, {49914, 1, 243, 72}, " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJK"}, {:ok, 'www.google.com.au', {58, 162, 62, 84}, {58, 162, 62, 84}, {49914, 0, 60, 65}, " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJK"}]

... hold that thought ...

if I follow the _procket_ readme.md instructions to setuid by running:

`sudo chmod 4750 deps/procket/priv/procket`

then executing as non privileged user seems to work OK.  Now all I have to figure out is:
- does escript package up the procket functionality with appropriate permissions set?
- is it possible to change the settings in my app so that when the repo is cloned, running `mix deps.get` asks for superuser permission prior to trying to compile the dependencies?
- is it possible to have `mix deps.get` or `mix deps.compile` run a `chmod` to allow for setuid on procket?

madlep commented on lib/quoil/cli.ex:
```diff
@@ -0,0 +1,61 @@
+    if exit_code != 0 do
+      IO.puts "ERROR: The ping command exited with code: #{exit_code}"
+      IO.puts rslt
+      System.halt(:abort)
```
if every ping fails, `ping` will return a status code of `2`. I'm guessing you probably don't want the app to stop testing pings if it hits a spell where it's having difficulties that may resolve between sets?

rawdamedia replied:
That depends... I just ran quoil on the Telstra DNS but was on my work wifi (i.e. not a on a Telstra network), so I got an error stating that it was rejected due to firewall rules... It would therefore have been futile to keep repeatedly attempting to run these pings and would be better to fail earlier.

I will have to think about how to handle this, because having 100% packet loss may or may not be a valid data point.

I guess that if I use gen_icmp, then I have more of a chance to finesse what happens after each individual ping, rather than waiting for the whole run to complete...

madlep commented on lib/quoil/repeat.ex:
```diff
@@ -0,0 +1,15 @@
+  def repeat_pings({nr, rpt_interval, args}) do
+    Stream.interval(rpt_interval * 60 * 1000)
+    |> Stream.take(nr)
+    |> Enum.map(fn _ -> Quoil.CLI.run_ping(args)|>Quoil.ParseResults.parse_result|>Quoil.LogResults.write_log end)
```
We're repeating the main app logic here.

I'd look at extracting the `repeat` and `wait` switches in `Quoil.CLI.main`, and doing this logic there.

madlep commented:
Overall, this is really good. Good uses of functional logic and transforming and extracting immutable elixir data.

:+1:

Commit:
removed administracja@miniserv.rawdanowicz.net.au

Commit:
new executable

Commit:
new executable

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
extract run_ping to new module

Commit:
gen_icmp for ping

pulled in gen_icmp as dependency
added —system command-line switch

Commit:
logging of errors added

Commit:
BUGFIX

Commit:
testing gen_icmp

can’t seem to get gen_icmp working - don’t know if it is a mac specific
problem -> will try linux

Commit:
gen_icmp - test whether able to be used

Commit:
BUGFIX for Ubuntu

fixed regex so that it also works on Ubuntu (slightly different output
of the system ping command)

Commit:
simplify specifying default switches

Commit:
Removed generated test coverage html files

Commit:
added comments

Commit:
fixed documentation

Commit:
updated compiled binary
